sopa/patches/infer.py

```diff
@@ -72,38 +74,44 @@ def _get_extraction_parameters(
     resize_factor = slide_metadata["level_downsamples"][level] / downsample
     patch_width = int(patch_width * downsample)
 
-    return level, resize_factor, patch_width, True
+    return level, resize_factor, patch_width, downsample, True
 
 
-class Embedder:
+class Inference:
     def __init__(
         self,
         image: MultiscaleSpatialImage | SpatialImage,
-        model_name: str,
+        model: Callable | str,
         patch_width: int,
         level: int | None = 0,
         magnification: int | None = None,
         device: str = "cpu",
     ):
         self.image = image
-        self.model_name = model_name
+
+        if isinstance(model, str):
+            assert hasattr(
+                models, model
+            ), f"'{model}' is not a valid model name under `sopa.embedding.models`. Valid names are: {', '.join(models.__all__)}"
+            self.model_str = model
+            self.model: torch.nn.Module = getattr(models, model)()
+        else:
+            self.model: torch.nn.Module = model
+
         self.device = device
 
         self.cs = get_intrinsic_cs(None, image)
 
         slide_metadata = image.attrs.get("metadata", {})
-        self.level, self.resize_factor, self.patch_width, success = _get_extraction_parameters(
-            slide_metadata, patch_width, level, magnification, backend=image.attrs["backend"]
+        self.level, self.resize_factor, self.patch_width, self.downsample, success = (
+            _get_extraction_parameters(
+                slide_metadata, patch_width, level, magnification, backend=image.attrs["backend"]
```

Comment:
For the WSI use cases it will always be set since but I integrated your suggestion anyway.

Reply:
Actually, for some spatial transcriptomics data, we may have some H&E data that comes with it
But it doesn't have the "usual" metadata, so it would not be possible to use these functions without this `.get` :)